Modifies `EntityStateStore` application to include support for dead-reckoning

met4000 opened the pull request:
Modifies `EntityStateStore` to use the new `DrEntityStatePdu`s, allowing dead-reckoning to be performed on stored entities.
Provides support for all 9 dead-reckoning algorithms in the DIS v7 specification.

Resolves: #93

met4000 committed:
Updates utils maths classes (Quaternion, Vec3)

Adds a number of methods to the Quaternion and Vec3 utility classes.
Notably adds support for Quaternion -> EulerAngles conversion, rotating
Vec3s by Quaternions, and introduces tests for Quaternion functionality.

met4000 committed:
Implements EntityStateStore's getUpdatedSince

met4000 committed:
Adds dead-reckoning support

Modifies EntityStateStore to provide dead-reckoning models for entities,
providing support for querying the entity's estimated position.

See: #93

michaelrfraser commented on codebase/src/java/disco/org/openlvc/disco/application/DeadReckoningModelStore.java:
```diff
@@ -0,0 +1,208 @@
+ * Tracks objects and provides dead-reckoning models based on their last known state.
+ * As updates are received the previous model is replaced.
+ */
+public class DeadReckoningModelStore implements IDeleteReaperManaged
```
Wondering whether this is similar enough to be merged with the EntityStore? If so you could just provide the following methods:

```
boolean isDeadReckoningAvailable( EntityId id );
DrmState getDeadReckonedStateAt( EntityId id, long localTimestamp );
DrmState getDeadReckonedStateAfter( EntityId id, double dt );
```

That would mean that all the entity related things are in one place, which makes things a bit simpler. Also dead reckoning calculations are only performed if you call the `getDeadReckonedStateXXX` methods, so its no additional overhead if the user don't want to utilize it.

michaelrfraser replied:
Had a chat to Nathan about this oob, and I we'll leave it like this for now. Argument for the current approach is that you can potentially DR more things than just entities, so planning ahead for the future it makes sense to have it on its own.

met4000 replied:
Marking as resolved for now

michaelrfraser commented on codebase/src/java/disco/org/openlvc/disco/application/DeadReckoningModelStore.java:
```diff
@@ -0,0 +1,208 @@
+	}
+
+	/**
+	 * @param entityId
```
Make sure your javadoc is complete here.

michaelrfraser commented on codebase/src/java/disco/org/openlvc/disco/application/deadreckoning/DeadReckoningModel.java:
```diff
@@ -0,0 +1,147 @@
+		DeadReckoningParameter params = object.getDeadReckoningParams();
+		switch( params.getDeadReckoningAlgorithm() )
+		{
+			case Other:
```
I think it reads better if the `default` case block is at the end. You can read it as "if all else fails, this is what happens". I'd keep the Other and Static cases with it as well.

michaelrfraser commented on codebase/src/java/disco/org/openlvc/disco/application/deadreckoning/model/FVW.java:
```diff
@@ -0,0 +1,86 @@
+
+		// displacement
+		Vec3 aDisp = new Vec3( dv );
+		aDisp.multiply(dt / 2.0);
```
Spaces needed between outer parentheses here. E.g.

```
aDisp.multiply( dt / 2.0 );
```

michaelrfraser commented on codebase/src/java/disco/org/openlvc/disco/utils/Quaternion.java:
```diff
@@ -72,19 +76,39 @@ public Quaternion multiply( Quaternion q )
+		{
+			// yaw and roll are aligned - use only yaw
+			yaw = yaw - roll;
+			pitch = pitchRatio > 0d ? EulerAngles.THETA_MAX : EulerAngles.THETA_MIN;
```
This is purely up to taste, but I think formatting tertiary operators the following way makes them much more readable.

```
pitch = pitchRatio > 0.0 ? EulerAngles.THETA_MAX
                         : EulerAngles.THETA_MIN;
```

michaelrfraser commented on codebase/src/java/disco/org/openlvc/disco/utils/Quaternion.java:
```diff
@@ -72,19 +76,39 @@ public Quaternion multiply( Quaternion q )
+		}
 		
-		return new EulerAngles( (float)yaw, (float)pitch, (float)(roll-_180DEGREES));
+		return new EulerAngles( (float)yaw, (float)pitch, (float)roll);
```
Space needed before outer parenthesis here. E.g.

```
return new EulerAngles( (float)yaw, (float)pitch, (float)roll );
```

michaelrfraser reviewed (changes requested):
Hey Nathan, just had a re-read of your latest changes and have a few more comments (again mostly documentation and logging related).

Might be also worth tagging in @timpokorny as this is getting closer to completion.

met4000 commented:
WIP/TODO from discussion:
- [x] allow for overriding the algorithm when making dr request to `DrEntityStatePdu` (including caching alongside timestamp)
- [ ] flag/config to disable dead-reckoning (make all algorithms behave like Static)
- [ ] body coordinate DR alg implementations
  - [ ] basic test for circular path (should return to original position)
- [ ] coordinate system handling
  - [ ] `DeadReckoningAlgorithm` support for testing what coordinate system an algorithm uses - new enum?
  - [ ] methods for switching `DrmState` between coordinate systems (static on `DrmState`?)
  - [ ] switch between coordinate systems if a `DrEntityStatePdu` algorithm override is of a different coordinate system

met4000 commented:
Not sure where application flags such as the one to disable dead-reckoning should be added - `DiscoConfiguration` as a `Flag` (description says `Flag`s are 'PDU parsing/processing flags')? `DiscoConfiguration` as a system property? Some new config handler on `DisApplication`? Currently going with system property loaded into `DiscoConfiguration`.

met4000 committed:
Loads DisApplication config before services

DisApplication services and helpers used to be loaded before the config
for the DisApplication was loaded. If they relied on config options then
they would only be initialised with the defaults, even if the config
used to initialise the DisApplication had been modified. This commit
reorders the DisApplication constructor to load the config first before
starting the services and helpers.

met4000 commented:
I'll still wait for Michael to give this a thorough look before merging, but @timpokorny interested to see if this new setup is in line with what you were thinking.

met4000 committed:
FIXUP - Formatting changes

timpokorny commented on codebase/src/java/disco/org/openlvc/disco/application/EntityStateStore.java:
```diff
@@ -52,6 +56,8 @@ protected EntityStateStore( PduStore store )
 		this.byId = new ConcurrentHashMap<>();
 		this.byMarking = new ConcurrentHashMap<>();
+
+		this.isDrEnabled = store.app.getConfiguration()::getDeadReckoningEnabled;
```
Very minor note, but I'd be inclined to make two small changes:
 - rename `store` to `parentStore` (as that makes the relationship between the down-chain EntityStateStore and the PduStore clear in the name (I told you it was minor)
 - I'd avoid the Boolean Supplier. Just store the parent and query it live. Generally, unless there is a great reason, in cases like this where we are creating connective tissue between components, I'd go for the whole object, rather than just one small cross-section of it (if that makes sense). It makes the connection more explicit.

timpokorny commented on codebase/src/java/disco/org/openlvc/disco/application/pdu/DrEntityStatePdu.java:
```diff
@@ -0,0 +1,333 @@
+	 * 
+	 * @return the {@link DeadReckoningAlgorithm} for this {@link DrEntityStatePdu}
+	 */
+	public DeadReckoningAlgorithm getDeadReckoningAlgorithm()
```
Suggest renaming to `getDefaultDeadReckoningAlgorithm`

timpokorny commented on codebase/src/java/disco/org/openlvc/disco/application/utils/DrmState.java:
```diff
@@ -0,0 +1,181 @@
+	//----------------------------------------------------------
+	//                      CONSTRUCTORS
+	//----------------------------------------------------------
+	public DrmState( WorldCoordinate location, VectorRecord linearVelocity, VectorRecord linearAcceleration, EulerAngles orientation, AngularVelocityVector angularVelocity )
```
Same comment as in other places on line wrapping at 100 char width unless it makes things really ugly. I'll stop making it now, but check the whole commit for me.

timpokorny reviewed (changes requested):
Some minor pieces to look at, but overall this is looking really good.

met4000 committed:
FIXUP - Formatting changes and suggested tweaks

met4000 committed:
FIXUP - Formatting tweaks in utils

met4000 committed:
FIXUP - Formatting changes and satisfying Sonarqube

met4000 commented:
Some minor modifications; the recent commits have been renamed to make their naming consistent and clearer how they will be rebased in the previous commits. The latest commit has new formatting fixes and code tweaks to make Sonarqube happy(er).
